Client/macos/MacOSBluetoothConnector.mm

```diff
@@ -0,0 +1,161 @@
+#include "MacOSBluetoothConnector.h"
+
+MacOSBluetoothConnector::MacOSBluetoothConnector()
+{
+
+}
+MacOSBluetoothConnector::~MacOSBluetoothConnector()
+{
+	//onclose event
+	if (isConnected()){
+		disconnect();
+	}
+}
+
+@interface AsyncCommDelegate : NSObject <IOBluetoothRFCOMMChannelDelegate> {
+	@public
+	MacOSBluetoothConnector* delegateCPP;
+}
+@end
+
+@implementation AsyncCommDelegate {
+}
+
+-(void)rfcommChannelOpenComplete:(IOBluetoothRFCOMMChannel *)rfcommChannel status:(IOReturn)error
+{
+	if ( error != kIOReturnSuccess ) {
+		fprintf(stderr,"Error - could not open the RFCOMM channel. Error code = %08x.\n",error);
+		return;
+	}
+	else{
+		fprintf(stderr,"Connected. Yeah!\n");
+	}
+}
+
+-(void)rfcommChannelData:(IOBluetoothRFCOMMChannel *)rfcommChannel data:(void *)dataPointer length:(size_t)dataLength
+{
+	delegateCPP->receivedBytes = (char*)dataPointer;
+	delegateCPP->receivedLength = dataLength;
+	delegateCPP->wantNewData = false;
+}
+
+
+@end
+
+int MacOSBluetoothConnector::send(char* buf, size_t length)
+{
+	int error = [(__bridge IOBluetoothRFCOMMChannel*)rfcommchannel writeSync:(void*)buf length:length];
+	if ( error != kIOReturnSuccess ) {
+		fprintf(stderr,"Error - could send bluetooth command. Error code = %08x.\n",error);
+	}
+	return length;
+}
+
+
+void MacOSBluetoothConnector::connectToMac(MacOSBluetoothConnector* macOSBluetoothConnector)
+{
+	macOSBluetoothConnector->running = 1;
+	//get device
+	IOBluetoothDevice *device = (__bridge IOBluetoothDevice *)macOSBluetoothConnector->rfcommDevice;
+	// create new channel
+	IOBluetoothRFCOMMChannel *channel = [[IOBluetoothRFCOMMChannel alloc] init];
+	// create sppServiceid
+	IOBluetoothSDPUUID *sppServiceUUID = [IOBluetoothSDPUUID uuidWithBytes:(void*)SERVICE_UUID_IN_BYTES length: 16];
+	// get sppServiceRecord
+	IOBluetoothSDPServiceRecord *sppServiceRecord = [device getServiceRecordForUUID:sppServiceUUID];
+	// get rfcommChannelID from sppServiceRecord
+	UInt8 rfcommChannelID;
+	[sppServiceRecord getRFCOMMChannelID:&rfcommChannelID];
+	// setup delegate
+	AsyncCommDelegate* asyncCommDelegate = [[AsyncCommDelegate alloc] init];
+	asyncCommDelegate->delegateCPP = macOSBluetoothConnector;
+	// try to open channel
+	if ( [device openRFCOMMChannelAsync:&channel withChannelID:rfcommChannelID delegate:asyncCommDelegate] != kIOReturnSuccess ) {
+		throw "Error - could not open the rfcomm.\n";
+	}
+	// store the channel
+	macOSBluetoothConnector->rfcommchannel = (__bridge void*) channel;
+
+	printf("Successfully connected");
+
+	// keep thread running
+	while (macOSBluetoothConnector->running) {
```

Comment:
Please don't use this approach. I know that we already talked about this, but inter thread signaling should not involve sleeps and unprotected bools. This way has two problems:
1. According to C++ memory rules, this can safely run forever. Because running is just a plain int, the compiler can assume that no one other this current loop changes it, and cache it in a register. That way, it'll NEVER know that the variable was changed to 0 from another thread. This can be solved by making it an atomic bool - `std::atomic<bool>`.
2. Sleeps should never be used for thread synchronization. Use a condition variable.

Reply:
I changed it to an atomic bool, but couldn't change it to a condition variable because that causes everything to freeze.
Is this good enough?

Reply:
Does macOS call the delegate in the same thread, like signals in Linux? very interesting